src/analysis_and_optimization/Mir_utils.ml

```diff
@@ -426,6 +426,388 @@ let rec update_expr_ad_levels autodiffable_variables
+ *      in the list of known failures the variable exists 
+ *      on the right hand side or in the indexed type?
+ *
+ * 4. (TODO: For loops)
+ *  Single indexing is fine outside of loops, but in loops we need to return true
+ *
+ * All other statements simply recurse into their lists of expressions 
+ *  or statements repeating the checks from above.
+ *)
+let rec query_aos_stmts (var_name : string)
+    (known_failures : string Set.Poly.t)
+    (stmt_map : (int, Stmt.Located.Non_recursive.t) Core_kernel.Map.Poly.t)
+    Stmt.Located.Non_recursive.({pattern; _}) : bool =
+  let query_expr = query_aos_exprs var_name known_failures in
+  let query_stmt = query_aos_stmts var_name known_failures stmt_map in
+  let query_name = query_aos_names var_name in
+  let get_key key = Map.find_exn stmt_map key in
+  let find_key blah = List.map ~f:get_key blah in
+  let find_pattern (pattern : (Expr.Typed.t, cf_state) Stmt.Fixed.Pattern.t) =
+    match pattern with
+    | Decl {decl_adtype= UnsizedType.DataOnly; decl_id; _}
+      when decl_id = var_name ->
+        true
+    | Decl {decl_id; decl_type= Type.Sized st; _}
+      when decl_id = var_name && any_not_eigen_type st ->
+        true
+    | Decl _ -> false
+    | Stmt.Fixed.Pattern.NRFunApp
+        ( StanLib
+            ((name : string), (_ : bool Fun_kind.suffix), Common.Helpers.SoA)
+        , (exprs : Expr.Typed.Meta.t Expr.Fixed.t list) ) -> (
+        let does_name_exist = List.exists ~f:query_name exprs in
+        match does_name_exist with
+        | false -> false
+        | true -> (
+          match name with
+          | x when Stan_math_signatures.is_reduce_sum_fn x -> true
+          | x when Stan_math_signatures.is_variadic_ode_fn x -> true
+          | _ ->
+              let check_fun_support =
+                Stan_math_signatures.query_stan_math_mem_pattern_support name
+                  (List.map ~f:find_args exprs)
+              in
+              (not check_fun_support) || List.exists ~f:query_expr exprs ) )
+    | NRFunApp (CompilerInternal _, _) -> false
+    | NRFunApp (UserDefined _, exprs) -> List.exists ~f:query_name exprs
+    | NRFunApp (StanLib (_, _, AoS), exprs) -> List.exists ~f:query_name exprs
+    | Assignment (((assign_name : string), (_ : UnsizedType.t), lhs), rhs) ->
+        let query_index = search_index query_expr in
+        let check_name item = assign_name = item in
+        let is_bad_assign = Set.Poly.exists ~f:check_name known_failures in
+        is_bad_assign || List.exists ~f:query_index lhs || query_expr rhs
+    | IfElse (predicate, true_stmt, op_false_stmt) ->
+        let pred_query = query_expr predicate in
+        let true_query = query_stmt (get_key true_stmt) in
+        let false_query =
+          match op_false_stmt with
+          | Some stmt -> query_stmt (get_key stmt)
+          | None -> false
+        in
+        pred_query || true_query || false_query
+    | Block stmts -> List.exists ~f:query_stmt (find_key stmts)
```

Comment:
You shouldn't need to do any manual recursion, the monotone framework handles that for you. You should only look as shallowly as you can to recognize the patterns you need. So I think the only case where you need to look at more than one statement at a time is the for-loop rule, were you need to check how the variable is indexed inside the loop